// fs-web-server-scaffold/test/config.js
/**
 * Created by rain on 2015/11/6.
 */

'use strict';
/*jslint node:true*/
const path = require('path');
const os = require('os');
const dev = process.env.DEBUG || true;
const product = {
    port: 80,
    mws: [{ //K8s探针去鉴权
        opts: {
            exclude: [{ p: '/healthz', o: 'GET' }]
        }
    }],
    dc: {
        url: "postgres://postgres:postgres@10.8.30.21:5432/postgres",
        opts: {
            pool: {
                max: 20,
                min: 10,
                idle: 10000
            }
        },
        models: []
    },
    logger: {
        level: 'debug',
        json: false,
        filename: path.join(__dirname, 'log', 'runtime.txt'),
        colorize: true,
        maxsize: 1024 * 1024 * 5,
        rotationFormat: false,
        zippedArchive: true,
        maxFiles: 10,
        prettyPrint: true,
        label: '',
        timestamp: true,
        eol: os.EOL,
        tailable: true,
        depth: null,
        showLevel: true,
        maxRetries: 1
    }
};

const development = {
    port: 4000,
    mws: [
        {
            entry: require('fs-auth-kernel').entry,
            opts: { remMaxAge: 30, loginUrl: "/login", exclude: ["/login", { p: '/healthz', o: 'GET' }] }   //{'/healthz'}k8s探针用去掉鉴权
        },
        {
            entry: require('fs-auth-url').entry,
            opts: {
                exclude: [{ p: '/healthz', o: 'GET' }]//k8s探针用去掉鉴权
            }
        }
    ],
    dc: {
        url: "sqlite://sqlite:sqlite@localhost:5432/sqlite",
        opts: {
            storage: path.join(__dirname, 'test.db'),
            define: {
                "timestamps": false, "freezeTableName": true
            }
        },
        models: [require('fs-auth-kernel').models,
        require('fs-auth-url').models]
    },
    logger: product.logger
};
development.logger.filename = path.join(__dirname, 'log', 'development.txt');

module.exports = dev ? development : product;
